File: Groupe1/TP2/src/couleurs.c
#include <stdio.h>
#include <stdbool.h>

// Définition de la structure Couleur
struct Couleur {
    int r;
    int g;
    int b;
    int a;
};

// Fonction pour comparer deux couleurs
bool comparerCouleurs(struct Couleur c1, struct Couleur c2) {
    return c1.r == c2.r && c1.g == c2.g && c1.b == c2.b && c1.a == c2.a;
}

// Fonction principale
int main() {
    // Déclaration d'un tableau de 100 couleurs
    struct Couleur couleurs[100] = {
        {255, 0, 0, 255}, {0, 255, 0, 255}, {0, 0, 255, 255}, 
        {255, 0, 0, 255}, {255, 255, 0, 255}, {0, 255, 0, 255}, 
        {0, 0, 255, 255}, {255, 0, 0, 255}, {255, 255, 0, 255}, 
        {0, 0, 255, 255}, {255, 0, 0, 255}, {0, 255, 0, 255}, 
        {0, 0, 255, 255}, {255, 0, 0, 255}, {255, 255, 0, 255}, 
        {0, 255, 0, 255}, {0, 0, 255, 255}, {255, 0, 0, 255}, 
        {255, 255, 0, 255}, {0, 0, 255, 255},
    };
// Structure pour stocker les couleurs et les occurrences
    struct Couleur couleursDistinctes[100];
    int occurrences[100] = {0};
    int nbCouleursDistinctes = 0;

    // Parcours du tableau
    for (int i = 0; i < 100; i++) {
        bool trouve = false;

        // Vérifier si la couleur est déjà dans le tableau des couleurs distinctes
        for (int j = 0; j < nbCouleursDistinctes; j++) {
            if (comparerCouleurs(couleurs[i], couleursDistinctes[j])) {
                occurrences[j]++;
                trouve = true;
                break;
            }
        }

        // Si la couleur n'est pas trouvée, l'ajouter
        if (!trouve) {
            couleursDistinctes[nbCouleursDistinctes] = couleurs[i];
            occurrences[nbCouleursDistinctes] = 1;
            nbCouleursDistinctes++;
        }
    }

    // Affichage des couleurs distinctes et leurs occurrences
    printf("Couleurs distinctes et leurs occurrences :\n");
    for (int i = 0; i < nbCouleursDistinctes; i++) {
        printf("Couleur (R: %d, G: %d, B: %d, A: %d) - Occurrences: %d\n",
               couleursDistinctes[i].r, couleursDistinctes[i].g, couleursDistinctes[i].b, couleursDistinctes[i].a, occurrences[i]);
    }

    return 0;
}

File: Groupe1/TP2/src/couleurs.py
import random

class Couleur:
    def __init__(self, R, G, B, A):
        self.R = R  #Rouge
        self.G = G  #Verte
        self.B = B  #Bleue
        self.A = A  #Alpha

    def __eq__(self, other):
        # Comparer deux couleur
        return (self.R, self.G, self.B, self.A) == (other.R, other.G, other.B, other.A)

    def __repr__(self):
        # Représentation sous forme de chaîne pour afficher la couleur en décimal
        return f"couleurs r,g,b {self.R} {self.G} {self.B} {self.A}"


# Génération d'une liste de 100 couleurs aléatoires
tableau_couleurs = [Couleur(random.randint(0, 255), random.randint(0, 255), random.randint(0, 255), 255) for i in range(100)]

# Liste pour stocker les couleurs distinctes et leur nombre d'occurrences
occurrences_couleurs = []

# Parcourir la liste de couleurs et compter les occurrences
for couleur in tableau_couleurs:
    found = False  # Indicateur pour savoir si la couleur a été trouvée

    # Vérifier si la couleur est déjà dans la liste des couleurs distinctes
    for item in occurrences_couleurs:
        if item[0] == couleur:  # Comparer les couleurs
            item[1] += 1  # Incrémenter le compteur
            found = True  # La couleur a été trouvée
            
    # Si la couleur n'a pas été trouvée après la boucle
    if not found:
        occurrences_couleurs.append([couleur, 1])

# Affiche les couleurs distinctes et leur nombre d'occurrences en décimal
for couleur, occurrences in occurrences_couleurs:
    print(f"{couleur} : {occurrences}")
